loader/src/ui/nodes/TextAreaV2.cpp

```diff
@@ -0,0 +1,377 @@
+
+        // Correct the Y position to be relative to the container height
+        line.label->setPositionY(m_container->getContentHeight() - line.currentHeight + line.label->getScaledContentHeight());
+        m_container->addChild(line.label);
+    }
+}
+
+void geode::TextAreaV2::updateLineAlignment(CCLabelBMFont* line) {
+    switch (m_alignment) {
+        case Alignment::Left:
+            line->setAnchorPoint({ 0, 1 });
+            line->setPositionX(0);
+            break;
+        case Alignment::Center:
+            line->setAnchorPoint({ 0.5f, 1 });
+            line->setPositionX(m_container->getContentWidth() / 2);
+            break;
+        case Alignment::Right:
+            line->setAnchorPoint({ 1, 1 });
+            line->setPositionX(m_container->getContentWidth());
+            break;
+    }
+}
+
+void geode::TextAreaV2::addEllipsis(Line& line) {
+    if (!m_ellipsis || line.lineNumber == -1) {
+        return;
+    }
+
+    line.text = line.text.find_first_not_of(' ') == std::string::npos ?
+        "..." :
+        line.text + std::string("...").substr(std::min<size_t>(3, line.text.size() - line.text.find_last_not_of('.') - 1));
+
+    line.label->setString(line.text.c_str());
+
+    if (this->isWidthOverflowing(line.label)) {
+        const size_t lineSize = line.text.size();
+        const size_t overflow = this->getOverflowAmount(line.label, lineSize + 3);
+
+        line.overflow = line.text.substr(lineSize - overflow - 3, overflow) + line.overflow;
+
+        line.label->setString((line.text = line.text.substr(0, lineSize - overflow - 3) + "...").c_str());
+    }
+}
+
+bool geode::TextAreaV2::isWidthOverflowing(const CCLabelBMFont* line) {
+    return m_wrappingMode != WrappingMode::NoWrap && m_maxWidth >= 0 && line->getScaledContentWidth() + this->getTotalPaddingX() > m_maxWidth;
+}
+
+size_t geode::TextAreaV2::getOverflowAmount(CCLabelBMFont* line, const size_t lineSize) {
+    for (size_t overflow = 1; overflow < lineSize; overflow++) {
+        CCNode* character = cocos::getChild(line, lineSize - overflow);
+
+        if ((character->getPositionX() - character->getContentWidth() / 2) * m_textScale <= m_maxWidth) {
+            return overflow;
+        }
+    }
+
+    return 0;
+}
+
+geode::TextAreaV2::Line geode::TextAreaV2::createLine(const std::string& text, Line& previousLine) {
+    CCLabelBMFont* line = CCLabelBMFont::create(text.c_str(), m_font.c_str());
+
+    line->setScale(m_textScale);
+    line->setColor({ m_textColor.r, m_textColor.g, m_textColor.b });
+    line->setOpacity(m_textColor.a);
+
+    const size_t lineSize = text.size();
+    const float currentHeight = m_linePadding + previousLine.currentHeight + line->getScaledContentHeight();
+    const size_t overflow = this->isWidthOverflowing(line) ? this->getOverflowAmount(line, lineSize) : 0;
+    const size_t overflowStart = lineSize - overflow;
+    const std::string finalText = text.substr(0, overflowStart);
```

Comment:
Similarly in cases like strings `auto` is fair to avoid due to the multiple different string types